### ISSUE TITLE
`tar_terra_rast()` doesn't save units or varnames

### ISSUE THREAD

**Issue by Aariq** [njtierney/geotargets#58]
I suspect this is because those metadata are stored in a "sidecar file" when `filetype = "GeoTIFF"`, but I haven't investigated further.

``` r
library(targets)
tar_script({
    make_rast <- function() {
        x <- terra::rast(system.file("ex/elev.tif", package = "terra"))
        terra::units(x) <- "m"
        terra::varnames(x) <- "elev"
        x
    }
    list(
        geotargets::tar_terra_rast(
            rast,
            make_rast()
        )
    )
})
tar_make()
#> ▶ dispatched target rast
#> ● completed target rast [0.017 seconds]
#> ▶ ended pipeline [0.148 seconds]
tar_load(rast)
terra::units(rast)
#> [1] ""
terra::varnames(rast)
#> [1] "rast"
terra::names(rast)
#> [1] "elevation"
```

<sup>Created on 2024-04-25 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

**Comment by Aariq** [comment]
Yeah, looks like units are store in an aux.json file.  `varnames()` aren't read back in by `rast()` so I guess that one isn't our problem.  Related to #37

**Comment by Aariq** [comment]
I've tried a half dozen or so differente GDAL drivers and they all create an aux.json file when a SpatRaster has units. Ideas:
1) always zip files after writing and us vsizip to read back in
2) have `tar_terra_rast(my_map)` create two targets: `my_map_rast` and `my_map_meta`.  Not sure how reading this back in would work though

**Comment by Aariq** [comment]
Found this: https://github.com/rspatial/terra/blob/master/NEWS.md#new-18
aux.json file is written for all drivers except when using `writeCDF()`

**Comment by njtierney** [comment]
Thanks for this! I'll take a look later this week on my geotargets day.

**Comment by amart90** [comment]
Hello folks! Thank you for the work on this package. I've been interested in it for a while and am now finally getting a chance to use it. I noticed that `geotargets::tar_terra_rast()` seems to be stripping the time attribute of my raster and I'm thinking that it is probably related to this. Here is a reproducible example:

```r
library(targets)
tar_script({
  make_rast <- function() {
    x <- terra::rast(system.file("ex/elev.tif", package = "terra"))
    terra::time(x) <- as.Date("2010-01-01")
    x
  }
  list(
    geotargets::tar_terra_rast(
      rast,
      make_rast()
    )
  )
})

tar_make()
tar_load(rast)
terra::time(rast)
#> NA
```

**Comment by Aariq** [comment]
Thanks @amart90, I am pretty sure it is related because I think time gets stored in a "sidecar" file.  A workaround might be to use the layer names to store times instead (if they're not already being used for something).  Try replacing the `terra::time(x)` with `names(x)`.

**Comment by Aariq** [comment]
I think we will eventually (soon?) have an alternative workflow that keeps the sidecar files with metadata like the time attribute (WIP in #63), so I might reach out to you for feedback

**Comment by Aariq** [comment]
Just adding a note that it sounds like `varnames()` are not actually expected to be preserved when writing and reading a raster.  According to `?varnames`:

>Note that newly created or derived SpatRasters always have a single variable (data source), and therefore the variable names are lost when processing a multi-variable SpatRaster.

**Comment by Aariq** [comment]
Ok y'all, I think there are two solid options for dealing with this issue, but I'm not sure which is best, so I've done some benchmarking and summarized some pros and cons.  Rather than clog this issue up with a giant comment, here's the report: https://viz.datascience.arizona.edu/benchmarking-geotargets/

tl;dr: Archiving the files (main and sidecar) is more "invisible" to users and more likely to work with distributed workers, but slower. Using a separate cache for the files is faster and doesn't add any dependencies, but may not work with certain distributed worker setups.

Pretty stuck on this without some feedback so let me know what you think @njtierney, @brownag, @amart90.

**Comment by amart90** [comment]
> Pretty stuck on this without some feedback so let me know what you think @njtierney, @brownag, @amart90.

Thanks for including me on this, @Aariq - take or leave my thoughts. I appreciate the thorough write up - it was super useful. I like the cleaner cache version, but for me it is most important that this function "just works," so I think I would prefer the archive option.

To address the additional dependencies: I think we could get rid of the withr and fs dependencies pretty easily (see below) and based on your rational for using `zip::zip()` I think it is probably worthwhile to add.

```r
# Can replace `fs::path_dir()` with `base::dirname()`
identical(
  fs::path_dir("path/to/file.csv"),
  base::dirname("path/to/file.csv")
)
#> [1] TRUE

# Can replace `fs::path_file()` with `base::basename`
identical(
  fs::path_file("path/to/file.csv"),
  base::basename("path/to/file.csv")
)
#> [1] TRUE

# Can replace `withr::local_tempdir` with base::tempdir() and unlink(tempdir())
# This is basically what withr::local_tempdir() does under the hood andways:
# https://github.com/r-lib/withr/blob/v3.0.1/R/tempfile.R#L86-L99
tmp <- withr::local_tempdir()
# Can be replaced with:
tmp <- base::tempdir()
on.exit(unlink(tmp, recursive = TRUE))

# Can replace `fs::path()` with `base::file.path()` *path() assigns additional "fs_path" class
identical(
  unclass(fs::path("path/to", "file.csv")),
  base::file.path("path/to", "file.csv")
)
#> TRUE

# Can replace `fs::dir_ls()` with `base::list.files()` *dir_ls() adds an additional class and names and sorts vector
setdiff(
  fs::dir_ls(),
  base::list.files()
)
#> character(0)
```

In regard to the being slower that the cache option: for _me_ the difference is probably pretty inconsequential. I could conceive of situations where it would be better to be faster (e.g., handling _many_ small rasters), but I don't know that that particular use case consideration overrides the potential benefits of the archive approach.

**Comment by njtierney** [comment]
Hey @Aariq thanks for the very nice and comprehensive write up, that was super clean and slick. And thank you @brownag for your helpful comments!

I agree with @brownag - I think the archive/zip method is my preference. Keeping things so they "just work" I think is a nice "batteries included" approach to geotargets that means everything is as nice as possible for the user. There's also a possibility we can do some tricks with vsizip, @mdsumner and I might be able to cook something up.

In terms of dependencies, we actually get `withr` for free because targets imports tidyselect which imports withr. So we don't need to worry about that if we want the niceties of `withr`. But we don't get `fs` for free. Here's a little function I wrote to explain that effectively just wraps `pak`:

``` r
# on master/main branch
can_we_add_a_dep_for_free <- function(deps){
    
    local_deps <- pak::pkg_deps(".")
    
    are_deps_already_free <- deps %in% local_deps$ref
    
    names(are_deps_already_free) <- deps
    
    if (are_deps_already_free) {
        cli::cli_inform(
            message = c(
                "{.pkg {deps}} can be added for free:",
                "Explanation:")
        )
        print(pak::local_deps_explain(deps = deps))
    }
    
    if (!are_deps_already_free){
        cli::cli_inform(
            message = c(
                "{.pkg {deps}} can not be added for free :("
            )
        )
    }
    
}

can_we_add_a_dep_for_free("withr")
#> ℹ Loading metadata database
#> ✔ Loading metadata database ... done
#> 
#> withr can be added for free:
#> Explanation:
#> geotargets -> targets -> tidyselect -> withr
can_we_add_a_dep_for_free("fs")
#> fs can not be added for free :(
```

<sup>Created on 2024-10-28 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

<details style="margin-bottom:10px;">
<summary>
Session info
</summary>

``` r
sessioninfo::session_info()
#> ─ Session info ───────────────────────────────────────────────────────────────
#>  setting  value
#>  version  R version 4.4.1 Patched (2024-07-08 r86915)
#>  os       macOS Sonoma 14.5
#>  system   aarch64, darwin20
#>  ui       X11
#>  language (EN)
#>  collate  en_US.UTF-8
#>  ctype    en_US.UTF-8
#>  tz       Australia/Perth
#>  date     2024-10-28
#>  pandoc   3.2 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/tools/aarch64/ (via rmarkdown)
#> 
#> ─ Packages ───────────────────────────────────────────────────────────────────
#>  package     * version date (UTC) lib source
#>  cli           3.6.3   2024-06-21 [1] CRAN (R 4.4.0)
#>  digest        0.6.36  2024-06-23 [1] CRAN (R 4.4.0)
#>  evaluate      0.24.0  2024-06-10 [1] CRAN (R 4.4.0)
#>  fansi         1.0.6   2023-12-08 [1] CRAN (R 4.4.0)
#>  fastmap       1.2.0   2024-05-15 [1] CRAN (R 4.4.0)
#>  fs            1.6.4   2024-04-25 [1] CRAN (R 4.4.0)
#>  glue          1.7.0   2024-01-09 [1] CRAN (R 4.4.0)
#>  htmltools     0.5.8.1 2024-04-04 [1] CRAN (R 4.4.0)
#>  knitr         1.48    2024-07-07 [1] CRAN (R 4.4.0)
#>  lifecycle     1.0.4   2023-11-07 [1] CRAN (R 4.4.0)
#>  pak           0.7.2   2024-03-17 [1] CRAN (R 4.4.0)
#>  pillar        1.9.0   2023-03-22 [1] CRAN (R 4.4.0)
#>  reprex        2.1.1   2024-07-06 [1] CRAN (R 4.4.0)
#>  rlang         1.1.4   2024-06-04 [1] CRAN (R 4.4.0)
#>  rmarkdown     2.27    2024-05-17 [1] CRAN (R 4.4.0)
#>  rstudioapi    0.16.0  2024-03-24 [1] CRAN (R 4.4.0)
#>  sessioninfo   1.2.2   2021-12-06 [1] CRAN (R 4.4.0)
#>  utf8          1.2.4   2023-10-22 [1] CRAN (R 4.4.0)
#>  vctrs         0.6.5   2023-12-01 [1] CRAN (R 4.4.0)
#>  withr         3.0.1   2024-07-31 [1] CRAN (R 4.4.0)
#>  xfun          0.46    2024-07-18 [1] CRAN (R 4.4.0)
#>  yaml          2.3.10  2024-07-26 [1] CRAN (R 4.4.0)
#> 
#>  [1] /Users/nick/Library/R/arm64/4.4/library
#>  [2] /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/library
#> 
#> ──────────────────────────────────────────────────────────────────────────────
```

</details>

**Comment by Aariq** [comment]
Thanks for the feedback @njtierney.  Just want to clarify that the comments above were from @amart90, not @brownag (@brownag, if you have any opinions on this, I'd still be open to hearing them!).  I'll plan on making a PR that implements this zip file solution as an option this week. Because the solution will be fairly "invisible" to users, I'll probably call the argument something like `keep_metadata`. This PR will supersede #106 and #63 which can then eventually be closed.
